### PR TITLE
Workaround for CC7 devtoolset7

### PR DESCRIPTION
**Pull request by davidrohr**
@ironMann : This seems to be needed to compile in our GPU CI, which uses CC7 with devtoolset7 (we cannot use our GCC-ToolChain due to the ROCm RPMs). There seems to be a bug in that compiler version, and the symbols are missing when the template is not instantiated explicitly.
Could you merge this and bump in alidist, preferably asap?

@ktf : With this and AliceO2Group/AliceO2#2968 the new GPU CI will become green.

### CODE REVIEW

**Comment by ironMann** [comment]
@davidrohr This feels wrong...  I'm looking into CI logs at: https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/2968/3b886c0c8f9321839d426680c74a465a9791b9ff/build_O2_gpu/fullLog.txt
 
Have you perhaps built some dependencies with GCC 4.8? The missing symbol there refers to the old definition of sting which should not be required to link this. Since c++11 definition changed... See: https://godbolt.org/z/k9eJvG

Is there an actual compiler bug report referencing this?

**Comment by davidrohr** [comment]
@ironMann : Indeed, I think it is this problem. We are building all software with the devtoolset7 gcc, but of course it pulls in system packages of CC7, which were built with gcc 4.8.
So if I understand correctly, RedHat patched their devtoolset7 gcc to be compatible with the system packages, using the old definitions.
In fact, the devtoolset7 gcc is incompatible to our GCC-Toolbox. This is actually the reason I had to switch everything to devtoolset7 for the GPU CI. The ROCm RPMs are build with devtoolset7 and pull in related stuff, so using the GCC-Toolbox is leading to all weird kinds of memory corruption.
Long story short, this is a dependency hell, but now I got a working setup for the GPU CI with devtoolset7, and the only problem seems to be the missing basic_string destructor, for which this PR is a workaround.

If you have a better solution, I am very interested, but for now I'd like to fix the CI make the PRs green again asap. The alternative would be to not test O2Suite/DataDistribution in the GPU CI but only O2.

**Comment by ktf** [comment]
This could actually be a side-effect of LTO usage.

**Comment by davidrohr** [comment]
Indeed, but I wouldn't disable LTO for this reason.

**Comment by ktf** [comment]
Agreed, especially if the template trick does the job.

**Comment by davidrohr** [comment]
We should anyway switch to CC8 soon, then we can revert this :|

**Comment by ironMann** [comment]
@davidrohr I'll bump things if the RPM builder does not complain. Otherwise, I'm also happy with leaving this package out of that CI until I had time to understand this a bit better.

**Comment by ironMann** [comment]
Reading some more about this: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=82172#c12
the problem seems to be triggered by compiling c++17, using the old string implementation (like devtoolset-7 does), and gcc7 bugs (probably too early version to do C++17 properly). 
And best of all, it leads to runtime crashes...

@davidrohr  @ktf Is there a way for you to base your CI on devtoolset8? 

Also, don't we have this situation when specifying --always-prefer-system to aliBuild and devtoolset is available?

**Comment by davidrohr** [comment]
Yes, we have the same problem if a user uses --always-prefer-system with devtoolset7 available.

I think we cannot easily use devtoolset8. I'd rather wait for Cern CentOS 8. Hopefully we will have CC8 soon and then we have to wait for amd to support it  right now it seems not to work. I'll raise that point when I meet the engineers.

But in any case, my patch cannot break anything in a "sane" setup, or do I miss something? And I would also revert it asap when we can switch to a new CentOS. And for now, we need a solution, so either this patch or we restrict the CI.

Kind Regards
David Rohr

**Comment by ironMann** [comment]
@davidrohr  In this case I'm in favor of excluding this from the CI in question because I can't test for runtime errors and memory corruption which the link I mentioned above states. 

If the old way of using a GCC compiled on our own is not going away and it can still build the O2, please exclude DataDistribution until this can be sorted out and tested.

**Comment by davidrohr** [comment]
@ironMann : Just for the record: If the setup is not already broken, this patch cannot cause any harm, because it just makes sure the basic_string is instantiated.
If the setup is already broken, it might crash in any case. That is how I had to learn that ROCm doesn't work with our GCC-Toolchain

@ktf : Could you change the gpu-builder from O2Suite to O2 then and do the "update start". We can revert to the full CI once we have a CC8 GPU builder.

**Comment by ironMann** [comment]
@davidrohr @ktf I don't question the patch, but I do need to be sure this is going to work properly. Making the broken compiler/linker happy is the least of my worries.

The builder managed to progress to the end https://alijenkins.cern.ch/job/build-any-ib/8422/consoleFull , and I ran some test on my own in the meantime... 
So I'll push a new tag

**Comment by ironMann** [comment]
https://github.com/alisw/alidist/pull/2095
